Source/KSCrash/Source/KSCrash/Recording/Tools/BSG_KSMachHeaders.m

```diff
@@ -0,0 +1,173 @@
+//
+//  BSG_KSMachHeaders.m
+//  Bugsnag
+//
+//  Created by Robin Macharg on 04/05/2020.
+//  Copyright © 2020 Bugsnag. All rights reserved.
+//
+
+#import <mach-o/dyld.h>
+#import <dlfcn.h>
+#import <Foundation/Foundation.h>
+#import "BSG_KSDynamicLinker.h"
+#import "BSG_KSMachHeaders.h"
+
+BSG_Mach_Binary_Images *bsg_get_mach_binary_images() {
+    return &bsg_mach_binary_images;
+}
+
+/**
+ * Store a Mach binary image-excapsulating struct in a dynamic array.
+ * The array doubles on filling-up.  Typical sizes is expected to be in < 1000 (i.e. 2-3 doublings, at app start-up)
+ * This should be called in a threadsafe way; we lock against a simultaneous add and remove.
+ */
+void bsg_add_mach_binary_image(BSG_Mach_Binary_Image_Info element) {
+    
+    os_unfair_lock_lock(&bsg_mach_binary_images_access_lock);
+    
+    // Expand array if necessary
+    if (bsg_mach_binary_images.used == bsg_mach_binary_images.size) {
+        bsg_mach_binary_images.size *= 2;
```

Comment:
strictly speaking this is not always true - `realloc` could fail to find enough memory to allocate the entire requested value. The paranoid way of doing things would be to query the size using the returned pointer and store that as the value.

Reply:
While an OOM (at the small scales we're dealing with here) is indicative of bigger problems I agree with your paranoia.  I've opted for a simple approach whereby a failing `realloc` simply doesn't store the image info or adjust the array size.  I don't try smaller resizings either; better, I think, to lose the image info than introduce complex, potentially error prone, reallocation heuristics.  Happy to dig further if you disagree.